### PR TITLE
Add cached file listing plugin for Mac

### PR DESCRIPTION
**Pull request by atcuno**

### CODE REVIEW

File: volatility/framework/plugins/mac/list_files.py [hunk 1]
```diff
@@ -0,0 +1,187 @@
...
+           
+        return loop_vnodes
+
+    @classmethod
+    def _build_path(cls, vnodes, vnode_name, parent_offset):
+        path = [vnode_name]
+        
+        while parent_offset in vnodes:
+            parent_name, parent_offset, _ = vnodes[parent_offset]
+            if parent_offset == None:
```
Comment (ikelos) [inline volatility/framework/plugins/mac/list_files.py:150]: `is None` rather than `== None`

File: volatility/framework/plugins/mac/list_files.py [hunk 2]
```diff
@@ -0,0 +1,187 @@
...
+                parent_offset = 0
+            
+            path.insert(0, parent_name)
+
+        if len(path) > 1:
+            path = "/".join(path)
```
Comment (ikelos) [inline volatility/framework/plugins/mac/list_files.py:156]: Presumably paths are never absolute?  It doesn't look like there's a way for this to produce an absolute path unless the first element contains a `/` at the start...

File: volatility/framework/plugins/mac/list_files.py [hunk 3]
```diff
@@ -0,0 +1,187 @@
...
+    @classmethod
+    def list_files(cls,
+                   context: interfaces.context.ContextInterface,
+                   layer_name: str,
+                   darwin_symbols: str) -> \
+            Iterable[interfaces.objects.ObjectInterface]:
+
+        vnodes = cls._walk_mounts(context, layer_name, darwin_symbols)
+
+        dprint("got %d vnodes" % len(vnodes))
```
Comment (ikelos) [inline volatility/framework/plugins/mac/list_files.py:171]: Given the function shouldn't really stay in here, we either need to bulk the function up if a quick debugging mechanism is needed, or remove the calls to it too...

**Review by ikelos — COMMENTED** [review]
Cool, looks good.  Some of the comments are more questions than changes, I think I didn't have my head around all the private classmethods, but having gotten to the last method it all makes sense.  Docstrings and typing information is always appreciated, but the big thing that needs to change are the debugging print functions/statements.  After that, happy to merge it...  5:)

File: volatility/framework/plugins/mac/list_files.py [hunk 4]
```diff
@@ -0,0 +1,187 @@
...
+        return [
+            requirements.TranslationLayerRequirement(name = 'primary',
+                                                     description = 'Kernel Address Space',
+                                                     architectures = ["Intel32", "Intel64"]),
+            requirements.SymbolTableRequirement(name = "darwin", description = "Mac Kernel"),
+            requirements.PluginRequirement(name = 'mount', plugin = mount.Mount, version = (1, 0, 0)),
+    ]
+
+    @classmethod
+    def _vnode_name(cls, vnode):
```
[the remark this answers is not in the document]
Comment (atcuno) [reply volatility/framework/plugins/mac/list_files.py:48]: Fixed.

File: volatility/framework/plugins/mac/list_files.py [hunk 5]
```diff
@@ -0,0 +1,187 @@
...
+        # roots of mount points have special name handling
+        if vnode.v_flag & 1 == 1:
+            v_name = vnode.full_path()
+            print("found root mount at %x" % vnode.dereference().vol.offset)
```
[the remark this answers is not in the document]
Comment (atcuno) [reply volatility/framework/plugins/mac/list_files.py:52]: This was a debug statement that should have been removed. It is gone now.

File: volatility/framework/plugins/mac/list_files.py [hunk 6]
```diff
@@ -0,0 +1,187 @@
...
+from volatility.framework.interfaces import plugins
+from volatility.framework.renderers import format_hints
+from volatility.framework.symbols import mac
+from volatility.plugins.mac import mount
+
+vollog = logging.getLogger(__name__)
+
+import sys
+
+def cprint(msg):
```
[the remark this answers is not in the document]
Comment (atcuno) [reply volatility/framework/plugins/mac/list_files.py:20]: Removed.

File: volatility/framework/plugins/mac/list_files.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (atcuno) [reply volatility/framework/plugins/mac/list_files.py:48]: That worked. I added it

File: volatility/framework/plugins/mac/list_files.py (hunk 3 above)
Comment (atcuno) [reply volatility/framework/plugins/mac/list_files.py:171]: Done.

File: volatility/framework/plugins/mac/list_files.py (hunk 2 above)
Comment (atcuno) [reply volatility/framework/plugins/mac/list_files.py:156]: The first element should be / or its an orphan file

File: volatility/framework/plugins/mac/list_files.py (hunk 1 above)
Comment (atcuno) [reply volatility/framework/plugins/mac/list_files.py:150]: Done. thanks.

File: volatility/framework/plugins/mac/list_files.py [hunk 7]
```diff
@@ -0,0 +1,187 @@
...
+
+    @classmethod
+    def _add_vnode(cls, vnode, loop_vnodes):
+        key = vnode.dereference().vol.offset
+        added = False
+
+        if not key in loop_vnodes:
+            # We can't do anything with a no-name vnode
+            v_name = cls._vnode_name(vnode)
+            if v_name == None:
```
[the remark this answers is not in the document]
Comment (atcuno) [reply volatility/framework/plugins/mac/list_files.py:82]: Done. thanks.

File: volatility/framework/plugins/mac/list_files.py [hunk 8]
```diff
@@ -0,0 +1,187 @@
...
+        try:
+            parent = vnode.v_parent
+        except exceptions.InvalidAddressException:
+            pass
+
+        return parent
+
+    @classmethod
+    def _add_vnode(cls, vnode, loop_vnodes):
+        key = vnode.dereference().vol.offset
```
[the remark this answers is not in the document]
Comment (atcuno) [reply volatility/framework/plugins/mac/list_files.py:76]: Interesting. I had problems at first with not being explicit, but now I get the same results without them. I had a bunch of issues getting this plugin working so maybe I fixed whatever that issue was along the way.

File: volatility/framework/plugins/mac/list_files.py [hunk 9]
```diff
@@ -0,0 +1,187 @@
...
+
+        if not key in loop_vnodes:
+            # We can't do anything with a no-name vnode
+            v_name = cls._vnode_name(vnode)
+            if v_name == None:
+                return added
+
+            parent = cls._get_parent(vnode)
+            if parent:
+                parent_val = parent.dereference().vol.offset
```
[the remark this answers is not in the document]
Comment (atcuno) [reply volatility/framework/plugins/mac/list_files.py:87]: done.

File: volatility/framework/plugins/mac/list_files.py [hunk 10]
```diff
@@ -0,0 +1,187 @@
...
+                parent_val = None
+        
+            loop_vnodes[key] = (v_name, parent_val, vnode)
+
+            added = True
+
+        return added
+
+    @classmethod
+    def _walk_vnode(cls, vnode, loop_vnodes):
```
[the remark this answers is not in the document]
Comment (atcuno) [reply volatility/framework/plugins/mac/list_files.py:98]: It adds vnodes to loop_vnodes as it goes (if any valid / non-duplicate ones are found)

**Comment by atcuno** [comment]
@ikelos  I commented on each one and just pushed a new commit that addresses them.

File: volatility/framework/plugins/mac/list_files.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (ikelos) [reply volatility/framework/plugins/mac/list_files.py]: It's not, that makes perfect sense, was just after an explanation which you provided.  5:)

**Comment by ikelos** [comment]
Awesome, thanks for the changes, looks good!  5:)

**Pull request merged**